### PR TITLE
#131: add an option for disable using highest card to select friends

### PR DESCRIPTION
**Pull request by cai-jack**

### CODE REVIEW

**Comment by cai-jack** [comment]
@rbtying can u share your ci/cd setup and the scripts. the lint and formatting errors are causing me quite a few fixes. Thanks!

**Comment by rbtying** [comment]
@cai-jack they're defined here:

https://github.com/rbtying/shengji/blob/master/.github/workflows/ci-workflow.yaml

**Comment by cai-jack** [comment]
Thanks! what tool do u use?

**Comment by rbtying** [comment]
The frontend is just `yarn lint`, which will run `tslint`. The CI/CD is via GitHub Actions

**Comment by cai-jack** [comment]
got it. Thanks.

File: core/src/game_state.rs [hunk 1]
```diff
@@ -296,6 +311,22 @@ impl PropagatedState {
         }))
     }
 
+    pub fn set_friend_selection_policy(
```
Comment (rbtying) [inline core/src/game_state.rs:314]: Let's take `policy: FriendSelectionPolicy` directly -- serde should handle the translation for us

File: core/src/game_state.rs [hunk 2]
```diff
@@ -1088,6 +1119,21 @@ impl ExchangePhase {
                 if friend.initial_skip >= self.num_decks {
                     bail!("need to pick a card that exists!")
                 }
+
+                if let FriendSelectionPolicy::HighestCardNotAllowed =
+                    self.propagated.friend_selection_policy
+                {
+                    if (self.trump.number() == Number::Ace
```
Comment (rbtying) [inline core/src/game_state.rs:1126]: I'd prefer a match here?


```
match (self.trump.number(), friend.card.number()) {
   (Number::Ace, Some(Number::King)) | (_, Some(Number::Ace)) => bail!("you can't pick the highest card as your friend"),
    _ => (),
}
```

File: core/src/interactive.rs [hunk 3]
```diff
@@ -253,6 +257,7 @@ pub enum Message {
     SetChatLink(Option<String>),
     SetNumDecks(Option<usize>),
     SetKittySize(Option<usize>),
+    SetFriendSelectionPolicy(String),
```
Comment (rbtying) [inline core/src/interactive.rs:266]: You can take a `FriendSelectionPolicy` here and not deal with manual serde

File: core/src/interactive.rs [hunk 4]
```diff
@@ -316,6 +321,8 @@ impl BroadcastMessage {
             AdvancementPolicySet { policy: AdvancementPolicy::DefendPoints } => format!("{} required players to defend on points", n?),
             KittySizeSet { size: Some(size) } => format!("{} set the number of cards in the bottom to {}", n?, size),
             KittySizeSet { size: None } => format!("{} set the number of cards in the bottom to default", n?),
+            FriendSelectionPolicySet { policy: FriendSelectionPolicy::Unrestricted} => "Friend can be selected by any card".to_string(),
```
Comment (rbtying) [inline core/src/interactive.rs:330]: `cargo fmt`? (or, fix the spaces)

**Comment by cai-jack** [comment]
finally got thru the ci build. Will work on your check request. Thanks.

File: core/src/game_state.rs (hunk 2 above)
Comment (cai-jack) [reply core/src/game_state.rs:1126]: fixed

File: core/src/interactive.rs (hunk 3 above)
Comment (cai-jack) [reply core/src/interactive.rs]: done

File: core/src/game_state.rs (hunk 1 above)
Comment (cai-jack) [reply core/src/game_state.rs:314]: done

File: core/src/interactive.rs (hunk 4 above)
Comment (rbtying) [inline core/src/interactive.rs:324]: nit

"{} allowed any non-trump card to be selected as a friend"

File: core/src/interactive.rs [hunk 5]
```diff
@@ -316,6 +321,8 @@ impl BroadcastMessage {
...
+            FriendSelectionPolicySet { policy: FriendSelectionPolicy::HighestCardNotAllowed} => "Friend can not be selected by highest card".to_string(),
```
Comment (rbtying) [inline core/src/interactive.rs:325]: nit

"{} disallowed the highest non-trump card, as well as trump cards, from being selected as a friend"

File: frontend/src/Initialize.tsx [hunk 6]
```diff
@@ -300,6 +311,18 @@ const Initialize = (props: Props) => {
...
           </label>
         </div>
+        <div>
+          <label>
+            Friend Selection Restriction:{' '}
+            <select
+              value={props.state.propagated.friend_selection_policy}
+              onChange={setFriendSelectionPolicy}
+            >
+              <option value="Unrestricted">Unrestricted</option>
```
Comment (rbtying) [inline frontend/src/Initialize.tsx:321]: In the user-facing text, let's use "Non-trump cards"

File: frontend/src/Initialize.tsx [hunk 7]
```diff
@@ -300,6 +311,18 @@ const Initialize = (props: Props) => {
...
+              <option value="HighestCardNotAllowed">Highest card not allowed</option>
```
Comment (rbtying) [inline frontend/src/Initialize.tsx:322]: "Non-trump cards, except the highest"

File: core/src/interactive.rs (hunk 5 above)
Comment (cai-jack) [reply core/src/interactive.rs:325]: trump card selection is not an option after my previous fix. change to: 

"{} disallowed the highest non-trump card from being selected as a friend"

**Comment by cai-jack** [comment]
as you wish:)

**Comment by cai-jack** [comment]
can u restart the ci run? the failure was caused by crates.io not available. Thanks.

**Pull request merged**